### ISSUE TITLE
Optimizer Needs a caching system

### ISSUE THREAD

**Issue by russelljjarvis** [russelljjarvis/NeuronunitOpt#21]
Caching system should be modelled on previous NU cached results with hashing of dictionaries used for logic to compute if a fresh run is warranted.

Appropriate entry points to caching system are outlined here.
Only call 'run_ga' if the hash is unique:

Only perform the exhaustive-grid-search if ParameterSet is unique/not hashed.

See links anchored to commits below.
Also forgot to mention the BluePyOpt code implements a caching system that I never use:
https://github.com/russelljjarvis/neuronunit/blob/master/neuronunit/optimisation/algorithms.py#L324

Its much less useful to me than what you propose.

**Comment by rgerkin** [comment]
@russelljjarvis As I am just now getting to this, I think the line numbers above are out of date (master branch has changed since you wrote that).  Can you refresh the above links, using a specified commit instead of master, e.g. 

https://github.com/russelljjarvis/neuronunit/blob/afdb633d72312af83a624e1af7e39af90a48890d/neuronunit/optimisation/optimization_management.py#L128

That one is probably right but I'm not sure about the second one.  

I think GitHub automatically formats the links if they are specified commits, since the content is fixed.  I just pasted `https://github.com/russelljjarvis/neuronunit/blob/afdb633d72312af83a624e1af7e39af90a48890d/neuronunit/optimisation/optimization_management.py#L128` above (but without backticks) and it gave me the formatting you see.

**Comment by russelljjarvis** [comment]
Yes, you are right about the first location. Its now moved a bit again, but its the same function signature.

https://github.com/russelljjarvis/neuronunit/blob/a27e984e93a89460dc59d08063ae3b26aa50a53a/neuronunit/optimisation/optimization_management.py#L305-L310

https://github.com/russelljjarvis/neuronunit/blob/master/neuronunit/optimisation/optimization_management.py#L2684-L2685

I think you could make a seperate class method 

`_run_ga(self,args,kwargs)`, which is actually really my present `run_ga(self,args,kwargs)`
You could replace my current `run_ga` with a cache check condition, and a run if cache False control flow.

**Comment by rgerkin** [comment]
@russelljjarvis 
What is the difference between [this optimize](https://github.com/russelljjarvis/neuronunit_opt/blob/a27e984e93a89460dc59d08063ae3b26aa50a53a/neuronunit/optimisation/optimization_management.py#L341) and [this one](https://github.com/russelljjarvis/neuronunit_opt/blob/a27e984e93a89460dc59d08063ae3b26aa50a53a/neuronunit/optimisation/optimization_management.py#L305) (or more generally between the TSD and TSS classes).  I can work on caching without knowing this, but I also want to know this.

**Comment by rgerkin** [comment]
A hash for an optimization result should probably be based on at least these things:
-- States of the model that are invariant through the optimization run (fixed parameters, backends, etc.)
-- The tests used including their observations
-- Which parameters were allowed to vary, and with what bounds
-- Optimization details (mu, n_gen, selection methods, etc.)

But looking at the code in the various `optimization` methods, and in `run_ga`, I cannot tell where or how it accesses the model, so I can't find where I should check a hash of these model invariants.  Could you point me to where the model is access by the optimizer in these functions?

**Comment by russelljjarvis** [comment]
TSS inherits from SciUnit.TestSuite, it is unimplemented, as inheriting from TestSuite is problematic, it was only included as a template that you may be interested in completing. TSD is what is actually used (see notebook chapters[1-10]. Especially 10.

TSD uses OptMan. 
OptMan contains:
```
self.tests (contains observations `self.tests[0].observations` for example).
self.boundary_dict and self.ranges (aliased variable names, same idea sorry).
self.hold_constant = not varied
```
OptMan is general enough to include exhaustive search. Exhaustive search does not use MU and ,NGEN so OptMan contains an internal class DeapObject = SciUnitObject defined in file [optimisations.py](https://github.com/russelljjarvis/neuronunit_opt/blob/a27e984e93a89460dc59d08063ae3b26aa50a53a/neuronunit/optimisation/optimisations.py#L143-L147)
```
OM.DO.hc # hold constant
OptMan.DeapObject.boundary_dict
```

https://github.com/russelljjarvis/neuronunit_opt/blob/a27e984e93a89460dc59d08063ae3b26aa50a53a/neuronunit/optimisation/optimisations.py#L143-L147

**Comment by russelljjarvis** [comment]
This implementation (not yet integrated, is a little more transparent).
```python
def make_ga_DO(explore_edges, max_ngen, test, \
        free_params = None, hc = None,
        selection = None, MU = None, seed_pop = None, \
           backend = str('RAW'),protocol={'allen':False,'elephant':True}):
    ss = {}
    try:
        free_params.pop('dt')
    except:
        pass
    if 'Iext' in free_params:
        free_params.pop('Iext')
    for k in free_params:
        if not k in explore_edges.keys() and k not in str('Iext') and k not in str('dt'):
            ss[k] = explore_edges[str(free_params)]
        else:
            ss[k] = explore_edges[k]
    if type(MU) == type(None):
        MU = 2**len(list(free_params))
    max_ngen = int(np.floor(max_ngen))
    if not isinstance(test, Iterable):
        test = [test]
    from neuronunit.optimisation.optimisations import SciUnitOptimisation
    DO = SciUnitOptimisation(MU = MU, error_criterion = test,\
     boundary_dict = ss, backend = backend, hc = hc, \
                             selection = selection,protocol=protocol)

    if seed_pop is not None:
        # This is a re-run condition.
        DO.setnparams(nparams = len(free_params), boundary_dict = ss)

        DO.seed_pop = seed_pop
        DO.setup_deap()
        DO.error_length = len(test)

    return DO



class TSD(dict):
    def __init__(self,tests={},use_rheobase_score=False):
       self.DO = None

       super(TSD,self).__init__()

       self.update(tests)
       if 'name' in self.keys():
           self.cell_name = tests['name']
           self.pop('name',None)
       else:
           self.cell_name = 'simulated data'

       self.use_rheobase_score = use_rheobase_score
       self.elaborate_plots  = elaborate_plots
       self.backend = None


    def optimize(self,param_edges,backend=None,protocol={'allen': False, 'elephant': True},\
        MU=5,NGEN=5,free_params=None,seed_pop=None,hold_constant=None):
        if type(free_params) is type(None):
            free_params=param_edges.keys()
        #experimental_name = self.cell_name
        self.DO = make_ga_DO(param_edges, NGEN, self, free_params=free_params, \
                           backend=backend, MU = 8,  protocol=protocol,seed_pop = seed_pop, hc=hold_constant)
        self.DO.MU = MU
        self.DO.NGEN = NGEN
        ga_out = self.DO.run(max_ngen = max_ngen)
        ga_out['dtc_pop'] = dtc_pop
        self.backend = backend
        if not hasattr(ga_out['pf'][0],'dtc') and 'dtc_pop' not in ga_out.keys():
            _,dtc_pop = DO.OM.test_runner(ga_out['pf'],self.DO.OM.td,self.DO.OM.tests)
            ga_out['dtc_pop'] = dtc_pop
        if str(self.cell_name) not in str('simulated data'):
            ga_out = self.elaborate_plots(self,ga_out)


        from sciunit.scores.collections import ScoreMatrix#(pd.DataFrame, SciUnit, TestWeighted)

        ##
        # TODO populate a score table pass it back to DO.OM

        return ga_out, DO
```

**Comment by rgerkin** [comment]
OK, so I think the long-run goal would be for `TSD` to be subclass `sciunit.TestSuite` (as `TSS` did), provided we can work out using an actual `TestSuite` rather than a `dict`.  And since a lot of the content of `optimize` is neuroscience-independent, it could even by in `sciunit.TestSuite` itself.  Then some `neuronunit.TestSuite` could subclass that and fold in the stuff about rheobase, elephant, etc.  Having optimization powers in sciunit itself would be very powerful.  Does `OptMan` make any neuroscience assumptions, or simulator assumptions?  

Basically, we can think about separating the work you've done on SciUnit-based optimization as separate from the work you've done on optimizing neuroscience models with neuroscience tests.  This could be a neat way to produce two distinct but logically related papers.  

All this can wait until next year.
